src/migrator/async.ts

```diff
@@ -0,0 +1,304 @@
+								const durationSinceLastRun =
+									Date.now() - migrationState.last_run_time.getTime();
+								if (
+									(migrationState.is_backoff &&
+										durationSinceLastRun < initMigrationState.backoffDelayMS) ||
+									(!migrationState.is_backoff &&
+										durationSinceLastRun < initMigrationState.delayMS)
+								) {
+									// will still execute finally block where the migration lock is released.
+									return;
+								}
+							}
+							// set last run time and run counter only when backoff time sync between
+							// competing instances is in sync
+							migrationState.last_run_time = new Date();
+							migrationState.run_counter += 1;
+
+							let migratedRows = 0;
+							// here a separate transaction is needed as this migration may fail
+							// when it fails it would break the transaction for managing the migration status
+							await sbvrUtils.db.transaction(async (migrationTx) => {
+								migratedRows =
+									(await asyncRunnerMigratorFn?.(migrationTx)) ?? 0;
+							});
+
+							migrationState.migrated_rows += migratedRows;
+							if (migratedRows === 0) {
+								// when all rows have been catched up once we only catch up less frequently
+								migrationState.is_backoff = true;
+								// only store the first time when migrator converged to all data migrated
+								migrationState.converged_time ??= new Date();
+							} else {
+								// Only here for the case that after backoff more rows need to be caught up faster
+								// If rows have been updated recently we start the interval again with normal frequency
+								migrationState.is_backoff = false;
+							}
+						} catch (err: unknown) {
+							migrationState.error_counter++;
+							if (err instanceof Error) {
+								if (
+									migrationState.error_counter %
+										initMigrationState.error_threshold ===
+									0
+								) {
+									migrationState.last_error_message = `${err.name} ${err.message}`;
+									(sbvrUtils.api.migrations?.logger.error ?? console.error)(
+										`${key}: ${err.name} ${err.message}`,
+									);
+									migrationState.is_backoff = true;
+								}
+							} else {
+								(sbvrUtils.api.migrations?.logger.error ?? console.error)(
+									`async migration error unknown: ${key}: ${err}`,
+								);
+							}
+						} finally {
+							if (migrationState.is_backoff) {
+								asyncScheduler = setInterval(
+									asyncRunner,
+									initMigrationState.backoffDelayMS,
+								);
+							} else {
+								asyncScheduler = setInterval(
+									asyncRunner,
+									initMigrationState.delayMS,
+								);
+							}
+							// using finally as it will also run when return statement is called inside the try block
+							// either success or error release the lock
+							await updateMigrationStatus(tx, migrationState);
+						}
+					});
+				});
+			};
+
+			let asyncScheduler = setInterval(asyncRunner, initMigrationState.delayMS);
```

Comment:
I think this should be a `setTimeout` so that the delay between executions is based on when the previous one stopped rather than when it started. I also think it'll help with a lot of the manual `clearInterval`s that are necessary

Reply:
Oh right! Thanks for cathing this.